### PR TITLE
Implementation of QCommandLineParser to handle the command line options

### PR DESCRIPTION
**Pull request by carikaser**
Implementation of QCommandLineParser to handle the command line options. (However, I did not implement the handling of the subcommands or any validity checks for them, as was said in the issue discussion.)

**Description**
Adding QCommandLineParser to handle the the command line options.

**Related Issues**
This pull-request fixes issue #2884

### CODE REVIEW

**Comment by ad-daniel** [comment]
just for information, we have automatic tests that ensure the code is formatted correctly, so if you don't want to lose your mind waiting for the results you can run the test locally `python3 tests/test_sources.py` or if you use VS code or Atom you can also install plugins so that it automatically formats the code when saving. More instructions about the coding style and the extensions are available [here](https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style/)

**Comment by carikaser** [comment]
> just for information, we have automatic tests that ensure the code is formatted correctly, so if you don't want to lose your mind waiting for the results you can run the test locally `python3 tests/test_sources.py` or if you use VS code or Atom you can also install plugins so that it automatically formats the code when saving. More instructions about the coding style and the extensions are available [here](https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style/)

Thank you so much, I will check that out !

**Comment by omichel** [comment]
This PR has many unrelated formatting changes, which makes it hard to review.
Can you please revert these changes, so that we clearly see what your contribution is?

File: src/webots/gui/WbGuiApplication.cpp [hunk 1]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-        if (logArgument.contains(",")) {
-          QStringList argumentsList = logArgument.split(",");
-          WbPerformanceLog::createInstance(argumentsList[0], argumentsList[1].trimmed().toInt());
-        } else
-          WbPerformanceLog::createInstance(logArgument);
-
+    } else if (subCommand == "log-performance" && parser.isSet(logPerfomanceOption)) {
+      QStringList logArgument = parser.values(logPerfomanceOption);
+      if (logArgument[0].contains(",")) {
+        QStringList argumentsList = logArgument[0].split(",");
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:327]: We try to use `const` everywhere it is possible to enhance readability and make the code more robust:
```suggestion
        const QStringList argumentsList = logArgument[0].split(",");
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 2]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-      if (equalCharacterIndex != -1) {
-        serverArgument = arg.mid(equalCharacterIndex + 1);
-        // remove starting/trailing double quotes
-        if (serverArgument.startsWith('"'))
-          serverArgument = serverArgument.right(serverArgument.size() - 1);
-        if (serverArgument.endsWith('"'))
-          serverArgument = serverArgument.left(serverArgument.size() - 1);
+    } else if (subCommand == "stream" && parser.isSet(streamOption)) {
+      args.pop_front();
+      QStringList streamArgument = parser.values(streamOption);
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:309]:
```suggestion
      const QStringList streamArgument = parser.values(streamOption);
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 3]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+      args.pop_front();
       batch = true;
       WbMessageBox::disable();
-    } else if (arg.startsWith("--update-proto-cache")) {
-      QStringList items = arg.split('=');
-      if (items.size() > 1)
-        mTaskArguments.append(items[1]);
+    } else if (subCommand == "update-proto-cache" && parser.isSet(updateProtoCacheOption)) {
+      args.pop_front();
+      QStringList protoCacheUpdate = parser.values(updateProtoCacheOption);
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:295]:
```suggestion
      const QStringList protoCacheUpdate = parser.values(updateProtoCacheOption);
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 4]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+        cout << tr("The '--mode=stop' option is deprecated. Please use '--mode=pause' instead.").toUtf8().constData() << endl;
+        mStartupMode = WbSimulationState::PAUSE;
+      } else if (mode[0] == "pause")
+        mStartupMode = WbSimulationState::PAUSE;
+      else if (mode[0] == "realtime")
+        mStartupMode = WbSimulationState::REALTIME;
+      else if (mode[0] == "fast")
+        mStartupMode = WbSimulationState::FAST;
+      else if (mode[0] == "run") {
+        cout << "Warning: `run` mode is deprecated, falling back to `fast` mode" << endl;
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:275]: This is not consistent with other printed message: it should probably go through `stderr` and be translated.
```suggestion
        cout << "Warning: `run` mode is deprecated, falling back to `fast` mode" << endl;
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 5]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-      mStartupMode = WbSimulationState::REALTIME;
-    else if (arg == "--mode=fast")
-      mStartupMode = WbSimulationState::FAST;
-    else if (arg == "--mode=run") {
-      cout << "Warning: `run` mode is deprecated, falling back to `fast` mode" << endl;
-      mStartupMode = WbSimulationState::FAST;
-    } else if (arg == "--no-rendering")
+    } else if (subCommand == "mode" && parser.isSet(modeOption)) {
+      args.pop_front();
+      QStringList mode = parser.values(modeOption);
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:260]:
```suggestion
      const QStringList mode = parser.values(modeOption);
```

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you. It looks pretty good.

However, I have a few change requests.

Also, I believe we should not use `QCoreApplication::translate()`, but `tr` instead which is more convenient and used everywhere else in the source code of Webots. See documentation [here](https://doc.qt.io/qt-6/qcoreapplication.html#translate).

Finally, we should avoid using `fprintf(stderr, ...)` and prefer the C++ counterpart: `cerr <<`.

**Comment by omichel** [comment]
Also, the behavior on this branch is now different from before. If I type `webots --unknown-option`, I am getting a dialog box:

image

Whereas the behavior on the master branch is to display in the terminal:
```
webots: invalid option: '--unknown-option'
Try 'webots --help' for more information.
```

I believe we should not change the original behavior, which is more user friendly (getting the error in the terminal where the command was issued).

File: src/webots/gui/WbGuiApplication.cpp [hunk 6]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-          logArgument = logArgument.left(logArgument.size() - 1);
-
-        if (logArgument.contains(",")) {
-          QStringList argumentsList = logArgument.split(",");
-          WbPerformanceLog::createInstance(argumentsList[0], argumentsList[1].trimmed().toInt());
-        } else
-          WbPerformanceLog::createInstance(logArgument);
-
+    } else if (subCommand == "log-performance" && parser.isSet(logPerfomanceOption)) {
+      QStringList logArgument = parser.values(logPerfomanceOption);
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:325]:
```suggestion
      const QStringList logArgument = parser.values(logPerfomanceOption);
```

**Comment by carikaser** [comment]
> Also, the behavior on this branch is now different from before. If I type `webots --unknown-option`, I am getting a dialog box:
> 
> image
> 
> Whereas the behavior on the master branch is to display in the terminal:
> 
> ```
> webots: invalid option: '--unknown-option'
> Try 'webots --help' for more information.
> ```
> 
> I believe we should not change the original behavior, which is more user friendly (getting the error in the terminal where the command was issued).

Thank you for the review. However, concerning this comment, I think it is a particularity of the QCommandLineParser to open a dialog box to display the help or if an error occurs, is this a problem?

**Comment by omichel** [comment]
Can't we change this default behavior of QCommandLineParser ? Your code seems to print this "invalid option" message to stderr, but it doesn't appear to work. Can you check it?

**Comment by omichel** [comment]
The documentation says "On Windows, QCommandLineParser uses message boxes to display usage information and errors if no console window can be obtained.". However, when I launch Webots from a MSYS2 console, I don't get the error message in the console.

File: src/webots/gui/WbGuiApplication.cpp [hunk 7]
```diff
@@ -193,110 +194,164 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+    {"disable-logging", QCoreApplication::translate("main", "Disable logging")},
+    {"enable-logging", QCoreApplication::translate("main", "Enable logging")},
+    {"log-level", QCoreApplication::translate("main", "Log level")},
+    {"no-sandbox", QCoreApplication::translate("main", "No sandbox")},
+    {"single-process", QCoreApplication::translate("main", "Single Process")},
+    {"remote-debugging-port", QCoreApplication::translate("main", "Remote Debugging Port")},
+  });
+
+  parser.process(QCoreApplication::arguments());
+  parser.parse(QCoreApplication::arguments());
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:246]: Maybe you should check the return value of the `parse()` method and if false, you should print on stderr the error, which you can retrieve with the `errorText()` method. See example [here](https://doc.qt.io/qt-6/qcommandlineparser.html).

**Comment by omichel** [comment]
I just fixed the terminal problem in a87aad67432bbc7c6b6ed54ed92cfe024dc24ada.

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you for the changes.
However, I have a problem with this PR: the world file argument is ignored.
On the master branch: `webots projects/robots/gctronic/e-puck/worlds/e-puck.wbt` opens this world whereas on this branch, it ignores the world file argument and open the previously opened world.
Could you try to fix it?

File: src/webots/gui/WbGuiApplication.cpp [hunk 8]
```diff
@@ -305,7 +306,7 @@ void WbGuiApplication::parseArguments() {
       args.pop_front();
       const QStringList streamArgument = parser.values(streamOption);
       if (streamArgument.size() != 1) {
-        cerr << tr("webots: invalid option: '--log-performance': log file path is missing.").toUtf8().constData();
+        cerr << tr("webots: invalid option : '--log-performance': log file path is missing.").toUtf8().constData();
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:309]: In English, there is no space before a colon:
```suggestion
        cerr << tr("webots: invalid option: '--log-performance': log file path is missing.").toUtf8().constData();
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 9]
```diff
@@ -339,19 +340,25 @@ void WbGuiApplication::parseArguments() {
...
+      mTask = FAILURE;
+    }
+    if (!(path.at(0)).isEmpty()) {
+      mStartWorldName = path.at(0);
+      parser.clearPositionalArguments();
+    }
+  }
   if (stream && !batch)
-    cerr << tr("Warning: you should also use --batch (in addition to --stream) for production.").toUtf8().constData() << endl;
+    cerr << "Warning: you should also use --batch (in addition to --stream) for production." << endl;
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:361]: Why did you remove the translation here?

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by omichel — CHANGES_REQUESTED** [review]
When I type `webots --mode=` or `webots --mode=unknown`, it doesn't display any error and runs normally while it was displaying errors in this case on the master branch.

**Comment by omichel** [comment]
Also, the help message is not as good as before (less precise, incomplete, not very useful):
```
webots --help
$ webots --help
Usage: C:\msys64\home\Olivier\webots\msys64\mingw64\bin\webots-bin.exe [options]

Options:
  -?, -h, --help                   Displays help on commandline options.
  --help-all                       Displays help including Qt specific options.
  -v, --version                    Displays version information.
  --minimize                       Minimize the window
  --fullscreen                     View window in fullscreen
  --mode <speed>                   Specify speed of simulation (default is
                                   realtime)
  --convert                        Convert
  --no-rendering                   No rendering
  --sysinfo                        System information
  --batch                          Batch
  --update-proto-cache <task>      Update the PROTO cache
  --update-world                   Update world
  --enable-x3d-meta-file-export    Enable x3d meta file export
  --stream                         Stream
  --stdout                         stdout
  --stderr                         stderr
  --log-performance <logArgument>  Log performance
  --disable-logging                Disable logging
  --enable-logging                 Enable logging
  --log-level                      Log level
  --no-sandbox                     No sandbox
  --single-process                 Single Process
  --remote-debugging-port          Remote Debugging Port
```

Whereas it used to be:
```
$ webots --help
Usage: webots [options] [worldfile]

Options:

  --help
    Display this help message and exit.

  --version
    Display version information and exit.

  --sysinfo
    Display information about the system and exit.

  --mode=<mode>
    Choose the startup mode, overriding application preferences. The <mode>
    argument must be either pause, realtime or fast.

  --no-rendering
    Disable rendering in the main 3D view.

  --fullscreen
    Start Webots in fullscreen.

  --minimize
    Minimize the Webots window on startup.

  --batch
    Prevent Webots from creating blocking pop-up windows.

  --stdout
    Redirect the stdout of the controllers to the terminal.

  --stderr
    Redirect the stderr of the controllers to the terminal.

  --stream[="key[=value];..."]
    Start the Webots streaming server. Parameters may be
    given as an option:
      port=1234          - Start the streaming server on port 1234.
      mode=<x3d|mjpeg>   - Specify the streaming mode: x3d (default) or mjpeg.
      monitorActivity    - Print a dot '.' on stdout every 5 seconds.
      disableTextStreams - Disable the streaming of stdout and stderr.

  --log-performance=<file>[,<steps>]
    Measure the performance of Webots and log it in the file specified in the
    <file> argument. The optional <steps> argument is an integer value that
    specifies how many steps are logged. If the --sysinfo option is used, the
    system information is prepended into the log file.

  convert
    Convert a PROTO file to a URDF, WBO, or WRL file.

Please report any bug to https://cyberbotics.com/bug
```

The goal of this PR is not to change the help message or the behavior, but just to use the QCommandLineParser class.
We want to keep the behavior similar to what it was before.

File: src/webots/gui/WbGuiApplication.cpp [hunk 10]
```diff
@@ -193,117 +194,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-    else if (arg == "--mode=run") {
-      cout << "Warning: `run` mode is deprecated, falling back to `fast` mode" << endl;
-      mStartupMode = WbSimulationState::FAST;
-    } else if (arg == "--no-rendering")
+    } else if (subCommand == "mode" && parser.isSet(modeOption)) {
+      args.pop_front();
+      const QStringList mode = parser.values(modeOption);
+      if (mode.size() != 1 || mode[0] == "=") {
+        cerr << tr("Error: Must specify one mode argument.").toUtf8().constData() << endl;
+      }
```
Comment (omichel) [inline src/webots/gui/WbGuiApplication.cpp:287]: [coding style](https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style#cs360-single-if-while-or-for-statement-must-be-written-without-brackets-the-statement-part-should-be-put-on-a-separate-line-in-case-of-nested-block-keep-brackets) (here and at other places):
```suggestion
      if (mode.size() != 1 || mode[0] == "=")
        cerr << tr("Error: Must specify one mode argument.").toUtf8().constData() << endl;
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by omichel** [comment]
Thank you, the behavior is now very good. We are almost there.

I just have another requirement, I found the error messages a little bit inconsistent:
```
$ webots --mode=unknown
webots: invalid option: 'unknown'
Try 'webots --help' for more information.

$ webots --stream="unknown"
webots: unknown option: 'unknown' in --stream
Try 'webots --help' for more information.
Warning: you should also use --batch (in addition to --stream) for production.

$ webots --unknown-option
Webots: Unknown option 'unknown-option'.
```

Some comments:
1. Sometime Webots is spelled with an uppercase letter, sometimes not. I would prefer the lowercase version as it corresponds to the command line.
2. Sometimes the "Try 'webots --help' for more information." is displayed in case of error, sometimes not (note: it is always displayed on the master branch). I think it should always be displayed in case of error.
3. Sometime the option with an unknown value is displayed (like in `--stream`), sometimes not (like in `--mode`). This should be more consistent. I believe we should display an error message like "webots: invalid value 'blahblah' in --option_name".

Although some of these inconsistencies are also present on the master branch, this PR is a good opportunity to address them.

**Comment by omichel** [comment]
Oops, I actually meant that we should use the spelling "webots" only when we refer to the command line, not the software in general for which the "Webots" is indeed better... Sorry for any confusion...

**Comment by omichel** [comment]
I just fixed a number of problems and crashes depending on the command argument passed to webots.

File: src/webots/gui/WbGuiApplication.cpp [hunk 11]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+    AttachConsole(ATTACH_PARENT_PROCESS);
+#endif
 
-  const int size = args.size();
-  for (int i = 1; i < size; ++i) {
-    const QString &arg = args[i];
-    if (arg == "--minimize")
+  parser.parse(QCoreApplication::arguments());
+  const QStringList unknownOptions = parser.unknownOptionNames();
+  QStringList args = parser.optionNames();
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:271]:
```suggestion
  const QStringList args = parser.optionNames();
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 12]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+    cout << tr("webots: invalid option: '--%1'").arg(unknownOptions[0]).toUtf8().constData() << endl;
+    mTask = FAILURE;
+  }
+  while (!args.isEmpty()) {
+    const QString subCommand = args.first();
+    if (subCommand == "help" && parser.isSet(helpOption)) {
+      mTask = HELP;
+    } else if (subCommand == "version" && parser.isSet(versionOption)) {
+      mTask = VERSION;
+    } else if (subCommand == "minimize" && parser.isSet(minimizeOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:283]:
```suggestion
    if (subCommand == "help" && parser.isSet(helpOption))
      mTask = HELP;
    else if (subCommand == "version" && parser.isSet(versionOption))
      mTask = VERSION;
    else if (subCommand == "minimize" && parser.isSet(minimizeOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 13]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
       mShouldMinimize = true;
-    else if (arg == "--fullscreen")
+    } else if (subCommand == "fullscreen" && parser.isSet(fullscreenOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:285]:
```suggestion
    else if (subCommand == "fullscreen" && parser.isSet(fullscreenOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 14]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-      mStartupMode = WbSimulationState::PAUSE;
-    else if (arg == "--mode=realtime")
-      mStartupMode = WbSimulationState::REALTIME;
-    else if (arg == "--mode=fast")
-      mStartupMode = WbSimulationState::FAST;
-    else if (arg == "--mode=run") {
-      cout << "Warning: `run` mode is deprecated, falling back to `fast` mode" << endl;
-      mStartupMode = WbSimulationState::FAST;
-    } else if (arg == "--no-rendering")
+    } else if (subCommand == "mode" && parser.isSet("mode")) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:287]:
```suggestion
    else if (subCommand == "mode" && parser.isSet("mode")) {
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 15]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+        mStartupMode = WbSimulationState::FAST;
+      else if (mode[0] == "run") {
+        cerr << tr("Warning: `run` mode is deprecated, falling back to `fast` mode").toUtf8().constData() << endl;
+        tryHelp();
+        mStartupMode = WbSimulationState::FAST;
+      } else {
+        cout << tr("webots: invalid value '%1' in --mode").arg(mode[0]).toUtf8().constData() << endl;
+        mTask = FAILURE;
+      }
+    } else if (subCommand == "no-rendering" && parser.isSet(noRenderingOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:310]:
```suggestion
    } else if (subCommand == "no-rendering" && parser.isSet(noRenderingOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 16]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
       mShouldDoRendering = false;
-    else if (arg == "convert") {
+    } else if (subCommand == "convert" && parser.isSet("convert")) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:312]:
```suggestion
    else if (subCommand == "convert" && parser.isSet("convert")) {
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 17]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+      const QStringList fileToConvert = parser.values(convertOption);
+      mTaskArguments = fileToConvert;
       mTask = CONVERT;
-      mTaskArguments = args.mid(i);
-      break;
-    } else if (arg == "--help")
-      mTask = HELP;
-    else if (arg == "--sysinfo")
+    } else if (subCommand == "sysinfo" && parser.isSet(sysinfoOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:316]:
```suggestion
    } else if (subCommand == "sysinfo" && parser.isSet(sysinfoOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 18]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
       mTask = SYSINFO;
-    else if (arg == "--version")
-      mTask = VERSION;
-    else if (arg == "--batch") {
+    } else if (subCommand == "batch" && parser.isSet(batchOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:318]:
```suggestion
    else if (subCommand == "batch" && parser.isSet(batchOption)) {
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 19]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-        mTaskArguments.append(items[1]);
+    } else if (subCommand == "update-proto-cache" && parser.isSet("update-proto-cache")) {
+      const QStringList protoCacheUpdate = parser.values(updateProtoCacheOption);
+      if (protoCacheUpdate.size() == 1)
+        mTaskArguments.append(protoCacheUpdate[0]);
       else
         mTaskArguments.clear();
       mTask = UPDATE_PROTO_CACHE;
-    } else if (arg.startsWith("--update-world"))
+    } else if (subCommand == "update-world" && parser.isSet(updateWorldOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:328]:
```suggestion
    } else if (subCommand == "update-world" && parser.isSet(updateWorldOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 20]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
       mTask = UPDATE_WORLD;
-    else if (arg == "--enable-x3d-meta-file-export")
+    } else if (subCommand == "enable-x3d-meta-file-export" && parser.isSet(x3DMetaFileExportOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:330]:
```suggestion
    else if (subCommand == "enable-x3d-meta-file-export" && parser.isSet(x3DMetaFileExportOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 21]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-      QString serverArgument;
-      int equalCharacterIndex = arg.indexOf('=');
-      if (equalCharacterIndex != -1) {
-        serverArgument = arg.mid(equalCharacterIndex + 1);
-        // remove starting/trailing double quotes
-        if (serverArgument.startsWith('"'))
-          serverArgument = serverArgument.right(serverArgument.size() - 1);
-        if (serverArgument.endsWith('"'))
-          serverArgument = serverArgument.left(serverArgument.size() - 1);
+    } else if (subCommand == "stream" && parser.isSet("stream")) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:332]:
```suggestion
    else if (subCommand == "stream" && parser.isSet("stream")) {
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 22]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
+      if (streamArgument.isEmpty() || streamArgument[0].isEmpty()) {
+        cerr << tr("webots: value is missing in '--stream'.\n").toUtf8().constData();
+        mTask = FAILURE;
+      } else {
+        stream = true;
+        parseStreamArguments(streamArgument[0]);
       }
-      parseStreamArguments(serverArgument);
-    } else if (arg == "--stdout")
+    } else if (subCommand == "stdout" && parser.isSet(stdoutOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:341]:
```suggestion
    } else if (subCommand == "stdout" && parser.isSet(stdoutOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 23]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
       WbLog::enableStdOutRedirectToTerminal();
-    else if (arg == "--stderr")
+    } else if (subCommand == "stderr" && parser.isSet(stderrOption)) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:343]:
```suggestion
    else if (subCommand == "stderr" && parser.isSet(stderrOption))
```

File: src/webots/gui/WbGuiApplication.cpp [hunk 24]
```diff
@@ -191,119 +189,206 @@ void WbGuiApplication::parseStreamArguments(const QString &streamArguments) {
...
-        if (logArgument.endsWith('"'))
-          logArgument = logArgument.left(logArgument.size() - 1);
-
-        if (logArgument.contains(",")) {
-          QStringList argumentsList = logArgument.split(",");
-          WbPerformanceLog::createInstance(argumentsList[0], argumentsList[1].trimmed().toInt());
-        } else
-          WbPerformanceLog::createInstance(logArgument);
-
+    } else if (subCommand == "log-performance" && parser.isSet("log-performance")) {
```
Comment (ad-daniel) [inline src/webots/gui/WbGuiApplication.cpp:345]:
```suggestion
    else if (subCommand == "log-performance" && parser.isSet("log-performance")) {
```

**Review by ad-daniel — CHANGES_REQUESTED** [review]
Code-wise there were just a few one-liners to unbracket left. I couldn't suggest all the changes at once due to deleted lines, hopefully I didn't make a mistake in the process.

Behavior wise I've only noticed two differences:
- In the current branch `./webots projects/robots/niryo/ned/worlds/ned.wbt --stream` gives an error. Now it requires a value, which in theory should be optional
- Before `./webots convert projects/robots/niryo/ned/protos/Ned.proto` generated the urdf, now it gives an error. In this case I think we should ameliorate the description as well, it's not really clear how it's supposed to be used (since supposedly it converts in different formats), nor why only this command is `convert` instead of `--convert` (it was already like this)

**Comment by omichel** [comment]
I didn't yet fixed the `convert` problem, but I am now wondering what is the added value of this `QCommandLineParser` with respect to the previous implementation. It turns out that we don't use the `QCommandLineParser` goodies such as the automatic display of the help message and the code to parse the options doesn't look simpler than before. So, I am questioning the usefulness of this idea. Do you see any advantage of using `QCommandLineParser` in this case?

**Comment by carikaser** [comment]
> I didn't yet fixed the `convert` problem, but I am now wondering what is the added value of this `QCommandLineParser` with respect to the previous implementation. It turns out that we don't use the `QCommandLineParser` goodies such as the automatic display of the help message and the code to parse the options doesn't look simpler than before. So, I am questioning the usefulness of this idea. Do you see any advantage of using `QCommandLineParser` in this case?

I was thinking the same as you, as we don't use the main features of the `QCommandLineParser` such as the automatic help, version or error display, it seems that there would not be an added value compared to the previous code (and even makes the code longer).  Thus, I would also say that the way the parsing was done before was more optimal in order to get the desired results and using the `QCommandLineParser` here seems unnecessary.

**Comment by omichel** [comment]
Great, so I am closing this PR.

**Pull request closed without merging**